File: After/src/Logic/Customers/PurchasedMovie.cs
using System;
using Logic.Common;
using Logic.Movies;
using Newtonsoft.Json;

namespace Logic.Customers
{
 
    public class PurchasedMovie : Entity
    {
        [JsonProperty]
        public virtual Movie Movie { get; protected set; }
        //[JsonProperty(IsReference = true, PropertyName = "_customer", TypeNameHandling =TypeNameHandling.All)]
        [JsonProperty]
        private Customer _customer;
        //[JsonIgnore]
        public virtual Customer Customer { get {
                return this._customer;
            } }
        [JsonProperty]
        private decimal _price;
        public virtual Dollars Price
        {
            get => Dollars.Of(_price);
            protected set => _price = value;
        }
        [JsonProperty]
        public virtual DateTime PurchaseDate { get; protected set; }
        [JsonProperty]
        private DateTime? _expirationDate;
        public virtual ExpirationDate ExpirationDate
        {
            get => (ExpirationDate)_expirationDate;
            protected set => _expirationDate = value;
        }
        [JsonConstructor]
        protected PurchasedMovie()
        {
        }
       // [JsonConstructor]// used for deserialization
        internal PurchasedMovie(Movie movie, Customer customer, Dollars price, ExpirationDate expirationDate)
        {
            if (price == null || price.IsZero)
                throw new ArgumentException(nameof(price));
            if (expirationDate == null || expirationDate.IsExpired)
                throw new ArgumentException(nameof(expirationDate));

            Movie = movie ?? throw new ArgumentNullException(nameof(movie));
            this._customer = customer ?? throw new ArgumentNullException(nameof(customer));
            Price = price;
            ExpirationDate = expirationDate;
            PurchaseDate = DateTime.UtcNow;
        }
        //[JsonConstructor]// used for deserialization
        //private PurchasedMovie(Movie movie, Dollars price, ExpirationDate expirationDate)
        //{
        //    if (price == null || price.IsZero)
        //        throw new ArgumentException(nameof(price));
        //    if (expirationDate == null || expirationDate.IsExpired)
        //        throw new ArgumentException(nameof(expirationDate));

        //    Movie = movie ?? throw new ArgumentNullException(nameof(movie));
        //    // JSON.net only provides a null customer here not sure why
        //    //Customer = customer ?? throw new ArgumentNullException(nameof(customer));
        //    Price = price;
        //    ExpirationDate = expirationDate;
        //    PurchaseDate = DateTime.UtcNow;
        //}
    }
}

File: After/src/Api/clientsrc/app/logic/Movies/PurchasedMovie.ts
import { Entity } from '../common/Entity';
import { Movie } from './Movie';
import { Customer } from '../customers/Customer';
import { Dollars } from '../customers/Dollars';
import { Data } from '@angular/router/src/config';
import { ExpirationDate } from '../customers/ExpirationDate';

export class PurchasedMovie extends Entity {
    private _movie: Movie;
    private _customer: Customer;
    private _price: Dollars;
    private _expirationDate: ExpirationDate;
    private _purchaseDate: Date;

    constructor (movie: Movie, customer: Customer, price: Dollars, expDate: ExpirationDate) {
        super();
        if (price === null || price.IsZero) {
            throw new Error('Bad price');
        }
        if (expDate === null || expDate.IsExpired) {
            throw new Error('Bad exiration date');
        }
        if (!movie) {
            throw new Error('Null movie');
        }
        if (!customer) {
            throw new Error('Null customer');
        }
        this._price = price;
        this._expirationDate = expDate;
        this._movie = movie;
        this._customer = customer;
        this._purchaseDate = new Date();
    }
    get  Movie(): Movie {
        return this._movie;
    }
    get  Customer(): Customer {
        return this._customer;
    }
    get PurchaseDate(): Data {
        return this._purchaseDate;
    }
    get Price(): Dollars {
        return this._price;
    }
    get ExpirationDate(): ExpirationDate {
        return this._expirationDate;
    }

   
  
}
